src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -270,7 +273,12 @@ data Level m h = Level {
 -- | A merging run is either a single run, or some ongoing merge.
 data MergingRun m h =
     -- | A merging of multiple runs.
-    MergingRun !MergePolicyForLevel !NumRuns !(StrictMVar m (MergingRunState m h))
+    MergingRun
+      !MergePolicyForLevel
+      !NumRuns
+      !(PrimVar (PrimState m) Int)
+      -- ^ The number of currently /unspent/ credits
```

Comment:
> Any option you like? Do you want to change it here or should I do it as part of my PR later?

I'll probably add `Prim` instances even if it is a bit of boilerplate

>  Also, semantically it seems like it doesn't make to track unspent credits when a merge is already completed. This suggests that the counter should live in OngoingMerge.
> 
> Is the reason for it being here simply for performance, avoiding having to look into the MVar on the hot path? Might make sense to add a short comment explaining that.

This is largely explained by `Note [Credits]`: it's outside so that you don't have to acquire the more expensive MVar lock every time you supply credits. Instead, we use atomic operations on the `unspentCredits` variable for maximum concurrency. Putting the counter in `OngoingMerge` is not an option, i think. Semantically it's not too weird: spending unspent credits on a `CompletedMerge` is just a no-op, and spending credits on an `OngoingMerge` does some actual work

Reply:
Both points sound reasonable.

Reply:
Ah, something I forgot is that `fetchAddInt` and co are defined only for `Int`s. I tried to get around that using `coerce` on for example `PrimVar s TotalCredits`, but it doesn't work very well. So instead I've gone with the slightly less nice thing of having `newtype` around the `PrimVar`s